### PR TITLE
[API, Events] Danger API

### PR DESCRIPTION
**Pull request by Misfiy**

### CODE REVIEW

File: Exiled.Events/EventArgs/Player/ChangingDangerStateEventArgs.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        /// <inheritdoc />
+        public API.Features.Player Player { get; }
+
+        /// <summary>
+        /// Gets the <see cref="DangerStackBase"/>.
+        /// </summary>
+        public DangerStackBase Danger { get; }
+
+        /// <summary>
+        /// Gets the type of Danger changing.
```
Comment (Vladislav-CS) [inline Exiled.Events/EventArgs/Player/ChangingDangerStateEventArgs.cs:46]:
```suggestion
        /// Gets the <see cref="DangerType"/> of Danger changing.
```

File: Exiled.Events/EventArgs/Player/ChangingDangerStateEventArgs.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        /// </summary>
+        public DangerType Type { get; }
+
+        /// <summary>
+        /// Gets a value indicating whether or not the danger is being activated. If false it is ending.
+        /// </summary>
+        public bool IsActivating { get; }
```
Comment (Vladislav-CS) [inline Exiled.Events/EventArgs/Player/ChangingDangerStateEventArgs.cs:53]: Make a new property that does `IsEnding => !isActivating`

File: Exiled.Events/Handlers/Player.cs [hunk 3]
```diff
@@ -537,6 +542,12 @@ public class Player
         /// <param name="ev">The <see cref="BanningEventArgs"/> instance.</param>
         public static void OnBanning(BanningEventArgs ev) => Banning.InvokeSafely(ev);
 
+        /// <summary>
+        /// Called before a players danger state changes.
```
Comment (Vladislav-CS) [inline Exiled.Events/Handlers/Player.cs:546]:
```suggestion
        /// Called before a player's danger state changes.
```

File: Exiled.Events/Patches/Events/Player/ChangingDangerState.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    using CustomPlayerEffects.Danger;
+    using Exiled.API.Enums;
+    using Exiled.API.Features;
+    using Exiled.Events.Attributes;
+    using Exiled.Events.EventArgs.Player;
+    using HarmonyLib;
+
+    /// <summary>
+    /// Patch for when player changes danger.
+    /// </summary>
```
Comment (Vladislav-CS) [inline Exiled.Events/Patches/Events/Player/ChangingDangerState.cs:22]:
```suggestion
    Adds <see cref="ChangingDangerState"/> event
    /// </summary>
```

File: Exiled.Events/Patches/Events/Player/ChangingDangerState.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+            if (!dangerToType.TryGetValue(__instance.GetType(), out DangerType type))
+                throw new ArgumentException($"{__instance} does not have a DangerType");
+
+            if (value == __instance._isActive)
+                return false;
+
+            ChangingDangerStateEventArgs ev = new(Player.Get(__instance.Owner), __instance, type, value);
+            Handlers.Player.OnChangingDangerState(ev);
+
+            if (!ev.IsAllowed)
```
Comment (Vladislav-CS) [inline Exiled.Events/Patches/Events/Player/ChangingDangerState.cs:52]: `return ev.IsAllowed`

**Review by Vladislav-CS — CHANGES_REQUESTED** [review]

**Review by Vladislav-CS — APPROVED** [review]

**Review by VALERA771 — APPROVED** [review]

**Review by VALERA771 — APPROVED** [review]

**Pull request merged**